docs/ref/models/querysets.txt

```diff
@@ -264,6 +264,73 @@ control the name of the annotation::
 For an in-depth discussion of aggregation, see :doc:`the topic guide on
 Aggregation </topics/db/aggregation>`.
 
+
+.. currentmodule:: django.db.models.query
+
+.. class:: FilteredRelation(relation_name, condition=Q())
+
+    .. versionadded:: 2.0
+    .. attribute:: FilteredRelation.relation_name
+
+        The name of the field on which you'd like to filter the relation.
+    .. attribute:: FilteredRelation.condition
+
+        A :class:`~django.db.models.Q` object to control the filtering.
+
+Annotate supports also a new kind of argument that is an instance of
+:class:`.FilteredRelation`.
+
+This kind of parameter allows to pass filtering options inside the the ``ON``
+clause when a ``JOIN`` is performed.
+It does not act on the default relationship but on the given ``alias`` given
+as name of the parameter (``pizzas_vegetarian`` in example below).
+
+Example::
+
+    >>> from django.db.models import FilteredRelation
+    >>> Restaurant.objects.annotate(
+    ...    pizzas_vegetarian=FilteredRelation(
+    ...        'pizzas', condition=Q(pizzas__vegetarian=True),
+    ...    )).filter(pizzas_vegetarian__name__icontains='Mozzarella')
+
+This query will join the model Restaurant only to Pizzas that have vegetarian
+flag to True. This results into a performance improvement, noticeable if the
+join is performed through a large number of rows. Because the filtering from
+the ``WHERE`` clause will only operate on a subset of Pizzas (the vegetarian
+ones).
+
+.. admonition:: You need to be aware of some limitations
+
+    The :class:`.FilteredRelation` can not contain
+    conditions that span over deep relations.
+    The condition can only apply on fields that satisfy the constraint
+    ``Field.is_relation is False``.
+
+    Example::
+
+        >>> Restaurant.objects.annotate(
+        ...    pizzas_with_toppings_startswith_n=FilteredRelation(
+        ...        'pizzas__toppings',
+        ...        condition=Q(pizzas__toppings__name__startswith='n'),
+        ...     )).filter(pizzas_with_toppings_startswith_n__isnull=False)
+
+    Will raise ``ValueError`` "Filtered relation
+    'pizzas_with_toppings_startswith_n' cannot operate on foreign key
+    'pizzas__toppings'".
+
+.. admonition:: :meth:`~.QuerySet.prefetch_related()` is not yet supported
+
+    If you intend to prefetch related objects from the alias,
+    ``AttributeError`` will be raised.
+
+
+.. admonition:: :meth:`~.QuerySet.only()` is not yet supported
+
+    If you intend use :meth:`~.QuerySet.only()` using the filtered_relation
+    alias, ``AttributeError`` will be raised.
```

Comment:
I think for both this and the above message, you should remove the "yet", as it implies it will be, but it might never get implemented.

I also think you could reword a bit like "If you try to use (QuerySet.method) with a FilteredRelation alias, AttributeError will be raised"

Reply:
thank you

Reply:
What does the AttributeError say? Generally, we would try to raise a more helpful message saying that it's not supported.